File: network_apis/python/quickstep.py
"""quickstep.py: Accesses the network command line interface of the quickstep database."""

import grpc
import NetworkCli_pb2
import NetworkCli_pb2_grpc

import pandas as pd



def sql_to_table(query):
  """Runs an SQL query on the quickstep database.

  Runs the passed SQL query on the quickstep database set in quickstepIP and quickstepPort
  and returns the result in a panda data frame.

  Args:
      query: postgresql query string to pass to quickstep database.

  Returns:
      A panda data frame representation of the returned data.

  Raises:
      ValueError: The query execution returns an error.
  """
  responseText,error = sql_to_text(query)
  if "ERROR" in error:
    raise ValueError(error)

  # Return for empty or malformed responses  (For instance an INSERT or UPDATE)
  if (not responseText or responseText[0] != "+"):
    return pd.DataFrame()

  # Split by newline's and skip first [0], third [2] and last entry from formatting
  rows = responseText.split("\n")[1:-1]

  # Split first row by '|' to get column names.
  # Skip the first and last result because they are before the first '|' and after the last '|'
  columnNames = rows[0].split("|")[1:-1]
  columnNames = map(unicode.strip, columnNames)  # strip whitespace
  rows = rows[1:]

  data = {}
  for columnName in columnNames:
    data[columnName] = [];

  for row in rows:
    if row[0] == "|":  # Avoids bad lines such as the "Time" at the very bottom
      temp = row.split("|")[1:-1]
      temp = map(unicode.strip, temp)  # strip whitespace
      for ind, columnName in enumerate(columnNames):
        data[columnName].append(temp[ind]);

  return pd.DataFrame(data)

quickstepIP="localhost"
quickstepPort="3000"

def sql_to_text(queryString):
  """Runs an SQL query on the quickstep database.

  Runs the passed SQL query on the quickstep database set in quickstepIP and quickstepPort
  and returns the result string.

  Args:
      query: postgresql query string to pass to quickstep database.

  Returns:
      A string representation of the returned data and any errors returned from
      query execution.
  """
  channel = grpc.insecure_channel(quickstepIP+":"+quickstepPort)
  stub = NetworkCli_pb2_grpc.NetworkCliStub(channel)
  response = stub.SendQuery(NetworkCli_pb2.QueryRequest(query=queryString))
  return response.query_result, response.error_result

def tables():
  """Runs an SQL query on the quickstep database and returns a list of tables.

  Runs a query on the quickstep database set in quickstepIP and quickstepPort
  and returns a list of tables in the database.

  Args:
      none

  Returns:
      A list of table name strings for tables in the quickstep database.

  Raises:
      ValueError: The query execution returns an error.
  """

  responseText, error = sql_to_text("\\d")
  if "ERROR" in error:
    raise ValueError(error)

  rows = responseText.split("\n")[4:-2]

  tables=[];

  for row in rows:
    table = row.split("|")
    if len(table)>0:
      tableName = table[0].strip()
      type = table[1].strip()
      blocks = table[2].strip()
      tables.append({"name":tableName,"type":type,"blocks":blocks})
  return tables

def columns(table):
  """Runs an SQL query on the quickstep database for a given table name and returns a list of columns.

  Runs a query on the quickstep database set in quickstepIP and quickstepPort
  and returns a list of columns in the given table.

  Args:
      table: Table name string to return columns for.

  Returns:
      A list of column dicts with properties "name" and "type".

  Raises:
      ValueError: The query execution returns an error.
  """
  responseText, error = sql_to_text("\\d " + table)
  if "ERROR" in error:
    raise ValueError(error)

  rows = responseText.split("\n")[3:-1]

  columns = [];

  for row in rows:
    column = row.split("|")
    if len(column) > 0:
      columnName = column[0].strip()
      dataType = column[1].strip()
      columns.append({"name":columnName,"type":dataType})
  return columns

def rows(table, count="ALL"):
  """Runs an SQL query for a given table name and returns a panda data frame with the table's data.

  Runs a query on the quickstep database set in quickstepIP and quickstepPort
  and returns a panda data frame with the table's data.

  Args:
      table: Table name string to return data for.
      count: The maximum number of rows to return.

  Returns:
      A panda data frame representation of the returned data.

  Raises:
      ValueError: The query execution returns an error.
  """
  limitString=((" LIMIT "+str(count)) if count!="ALL" else "")
  return sql_to_table("SELECT * FROM " + table + " ORDER BY " + columns(table)[0]["name"] + limitString + ";")

File: network_apis/nodejs/quickstep.js
var grpc = require('grpc');
var quickstepGRPC = grpc.load("../../cli/NetworkCli.proto").quickstep;

var quickstep={
  ip:"localhost",
  port:"3000",

  sql_to_text:function(query,callback)
  {
    var client = new quickstepGRPC.NetworkCli(quickstep.ip+":"+quickstep.port,
                                         grpc.credentials.createInsecure());

    client.sendQuery({"query": query}, function(err, response) {
      if(response)
      {
        callback(response.query_result,response.error_result);
      }
      else {
        callback("","ERROR: No response received from quickstep database.");
      }
    });
  },

  sql_to_table:function(query,callback)
  {
    this.sql_to_text(query,function(result,error)
    {
      if(error.indexOf("ERROR")!=-1)
      {
        callback("",error);
        return;
      }

      if(!result || result[0]!="+")
      {
        return [];
      }

      var rows=result.split("\n");
      var columnNames=rows[1].split("|");
      columnNames=columnNames.map(Function.prototype.call,String.prototype.trim);

      var dataTable=[];
      console.log(result);
      for(var j=3;j<rows.length-3;j++)
      {
        var values=rows[j].split("|");
        dataTable[j-3]={};
        for(var k=1;k<columnNames.length-1;k++)
        {
          dataTable[j-3][columnNames[k].trim()]=values[k].trim();
        }
      }

      callback(dataTable,"");
    });
  },

  tables:function(callback)
  {
    this.sql_to_text("\\d",function(result,error)
    {
      if(error.indexOf("ERROR")!=-1)
      {
        callback("",error);
        return;
      }

      var rows=result.split("\n");
      var columnNames=["name","type","block"];
      var dataTable=[];
      console.log(result);
      for(var j=4;j<rows.length-2;j++)
      {
        var values=rows[j].split("|");
        dataTable[j-4]={};
        for(var k=0;k<columnNames.length;k++)
        {
          dataTable[j-4][columnNames[k].trim()]=values[k].trim();
        }
      }

      callback(dataTable,"");
    });
  },

  columns:function(table, callback)
  {
    this.sql_to_text("\\d "+table,function(result,error)
    {
      if(error.indexOf("ERROR")!=-1)
      {
        callback("",error);
        return;
      }

      var rows=result.split("\n");
      var columnNames=["name","type"];
      var dataTable=[];
      console.log(result);

      for(var j=3;j<rows.length-1;j++)
      {
        var values=rows[j].split("|");
        dataTable[j-3]={};
        for(var k=0;k<columnNames.length;k++)
        {
          dataTable[j-3][columnNames[k].trim()]=values[k].trim();
        }
      }

      callback(dataTable,"");
    });
  },

  rows:function(table,count="ALL",callback)
  {
    var limitString= count!="ALL" ? " LIMIT "+count : "";
    context=this;
    context.columns(table,function(columnData,error){
        if(error.indexOf("ERROR")!=-1)
        {
          callback("",error);
          return;
        }
        var sqlString="SELECT * FROM "+table+
          " ORDER BY "+columnData[0].name +
          limitString+";";
        context.sql_to_table(sqlString,function(rowData,error){
          if(error.indexOf("ERROR")!=-1)
          {
            callback("",error);
          }
          callback(rowData,"");
        });
    });
  }
};

module.exports=quickstep;
